src/picture.rs

```diff
@@ -103,15 +102,30 @@ pub unsafe extern "C" fn dav1d_default_picture_alloc(
     let y_sz = (y_stride * aligned_h as isize) as usize;
     let uv_sz = (uv_stride * (aligned_h >> ss_ver) as isize) as usize;
     let pic_size = y_sz + 2 * uv_sz;
-    let buf = rav1d_mem_pool_pop(
-        cookie as *mut Rav1dMemPool,
-        pic_size + RAV1D_PICTURE_ALIGNMENT - ::core::mem::size_of::<Rav1dMemPoolBuffer>(),
-    );
-    if buf.is_null() {
-        return Rav1dResult::<()>::Err(ENOMEM).into();
+
+    let pool = &*cookie.cast::<Arc<MemPool<u8>>>();
+    let pool = pool.clone();
+    let pic_cap = pic_size + 2 * RAV1D_PICTURE_ALIGNMENT;
+    // TODO fallible allocation
+    let mut buf = pool.pop(pic_cap);
+    if cfg!(debug_assertions) {
+        buf.resize_with(pic_cap, Default::default);
+    } else {
+        assert!(buf.capacity() >= pic_cap);
+        // SAFETY: We checked the capacity, so it is safe on that front.
+        // As for initialized, the bytes are not yet initialized.
+        // We do materialize intermediate slices to them, which is technically UB,
+        // but we do write to them eventually before any reads.
+        // This is done for performance, as zeroing causes a 10% performance regression.
+        buf.set_len(pic_cap);
```

Comment:
I was approaching this similarly to what we do for `DisjointMut`.  That is, with `cfg!(debug_assertions)` on, things should be fully safe, like for `DisjointMut`.

Reply:
`DisjointMut` isn't fully safe with debug checks enabled, it's performing all the same unsafe operation just with additional checks enabled to catch incorrect usage. Using `resize_with` in debug mode doesn't do anything to help catch incorrect usage, and I don't think having this subtle difference in behavior between debug and release builds is a good idea.

Reply:
The point isn't that `DisjointMut` is unsound in debug builds, it's that the only difference in behavior are additional debug checks. That's not what you have here, and I don't think having behavior differs between builds like this is good.

Reply:
Okay, I'll change it to have the same behavior always.  Now that I think about it more, since the allocator is overridable from C, everything may be uninitialized, so we should model things as `MaybeUninit`.  For this PR I will just do `set_len` unconditionally.

Reply:
I decided to make them `MaybeUninit<u8>`s instead and remove the `unsafe` `.set_len` entirely.  My thinking is that arbitrary C pic allocators won't initialize data anyways, so we need to model the `Rav1dPictureData::data` fields as maybe uninitialized anyways (though they're still raw pointers for now; I'm working on that next).  Does that sound okay?  I agree with your point that they should have the same behavior in debug builds, but realizing other allocators won't initialize the data anyways, I think this is better.

Reply:
Yeah, I think that's a better approach. Actually capturing that the buffer may be uninitialized seems more correct.